File: Editorial/129_Sum_Root_to_Leaf_Numbers/C/2_Recursive_Preorder_Traversal.c
// C symbolic Node structure for the Leetcode Playground environment
// struct TreeNode {
//     int val;
//     struct TreeNode *left;
//     struct TreeNode *right;
// };
void preorder(struct TreeNode* r, int currNumber, int* rootToLeaf) {
    if (r != NULL) {
        currNumber = currNumber * 10 + r->val;
        // if it's a leaf, update root-to-leaf sum
        if (r->left == NULL && r->right == NULL) *rootToLeaf += currNumber;
        preorder(r->left, currNumber, rootToLeaf);
        preorder(r->right, currNumber, rootToLeaf);
    }
}
int sumNumbers(struct TreeNode* root) {
    int rootToLeaf = 0;
    preorder(root, 0, &rootToLeaf);
    return rootToLeaf;
}

File: Editorial/129_Sum_Root_to_Leaf_Numbers/Go/2_Recursive_Preorder_Traversal.go
func sumNumbers(root *TreeNode) int {
    var preorder func(r *TreeNode, currNumber int)
    rootToLeaf := 0
    preorder = func(r *TreeNode, currNumber int) {
        if r != nil {
            currNumber = currNumber*10 + r.Val
            // if it's a leaf, update root-to-leaf sum
            if r.Left == nil && r.Right == nil {
                rootToLeaf += currNumber
            }
            preorder(r.Left, currNumber)
            preorder(r.Right, currNumber)
        }
    }
    preorder(root, 0)
    return rootToLeaf
}
